Drivers: Intel: DMIC: Fix FW panic with multiple DAI capture, pausing, stopping and resuming

singalsu opened the pull request.

kv2019i commented on src/drivers/intel/dmic.c:
```diff
@@ -1649,9 +1656,9 @@ static int dmic_remove(struct dai *dai)
+	if (uncached_dmic_active_fifos_mask || uncached_dmic_pause_mask)
 		return 0;
 
 	pm_runtime_put_sync(DMIC_CLK, dai->index);
```
This is really problematic. There's no actual reference counting in cavs_pm_runtime_dis_dmic_clk_gating() and the "dai->index" passed to pm_runtime_put_sync() is just printed out.

The problem is that if we switch to pm_runtime_put_sync() that actual does refcounting, this code will break. FYI @lgirdwood @lyakh @marc-hb 

Oh well, this goes beyond this PR's scope.

lgirdwood replied:
Lets get the ref counting in the PM code. This PM code should also be able to transfer to Zephyr with minimal refactoring.

lgirdwood replied:
it looks like we may need some locking around the mask manipulations .

singalsu replied:
I added 3rd commit about spinlocks. I was testing spinlock add in dmic_remove() but it froze. It turned out that dai_put() had already applied spinlock. I also found out that the registers write function was already covered by spinlock in set_config() so the patch adds only comments about spinlocks applied.

kv2019i reviewed (approved):
Some notes inline, but actual patch content looks good!

singalsu commented on src/drivers/intel/dmic.c:
```diff
@@ -136,7 +136,7 @@ static const uint32_t coef_base_b[4] = {PDM0_COEFFICIENT_B, PDM1_COEFFICIENT_B,
 /* Global configuration request for DMIC, need to use uncached address to access them */
 static SHARED_DATA struct sof_ipc_dai_dmic_params *dmic_prm[DMIC_HW_FIFOS];
-static SHARED_DATA int dmic_active_fifos_mask;
+static SHARED_DATA uint32_t dmic_active_fifos_mask;
```
Yep, that's the purpose, 16 bit int devices are not known in SOF plus new usage is unsigned. I added the new uint32_t pause_mask next to it so the active_fifos_mask looked weird. This was made a separe commit.

singalsu commented:
I just updated the PR: Added pause_mask clear to dmic_stop() in case it was called to stop without pause. A typo was fixed. Also added comments about spinlocks applied. Thanks for help with these @ranj063!

ranj063 commented:
looks good. Thanks @singalsu

lgirdwood reviewed (approved):
Lets fix the PM reference counting incrementally.
